proposals/cli_schema.md

```diff
@@ -0,0 +1,246 @@
+3. All file names must be checked for existence and validity.
+4. Objects in the `results[]` section should be validated against the rule type to determine if the necessary mandatory fields are present.
+5. Policies which are of type `Policy` must specify Namespace.
+6. There must not be duplicate names in resources.
+7. Deprecated fields must be handled with warnings to users.
+8. Breaking changes should cause the apiVersion to be incremented.
+
+```yaml
+apiVersion: cli.kyverno.io/v1beta1
+kind: KyvernoTest
+metadata:
+  name: test-check-nvidia-gpus
+  labels:
+    foolabel: foovalue
+  annotations:
+    fookey: foovalue
+spec:
+  policies:
+    # support fully-qualified file paths (https://github.com/kyverno/kyverno/issues/2315)
+    # wherever files are the value. Should be relative and absolute.
+  - check-nvidia-gpu.yaml
+  - relative/path/to/next/policy.yaml
+  - /fully/qualified/path/to/next/policy.yaml
+  resources:
+  - resource.yaml
+  - relative/path/to/next/resource.yaml
+  - /fully/qualified/path/to/next/resource.yaml
+  results:
+  # results[] vary depending on the rule type being tested.
+  # validate
+  - policy: policyname01 # required
+    rule: validaterulename # required
+    # resources must support multiple resources, all of which
+    # should be subject to the rest of the field values 
+    # in this result. (https://github.com/kyverno/kyverno/issues/2857)
+    resources: # required
+    - resource01
+    - resource02
+    - resource03
+    # namespace should set the namespace name for the resource(s), if not present,
+    # but otherwise should override the Namespace name if present. If the policy
+    # being tested is a `Policy` (Namespaced), and if the namespace is not defined
+    # in the resource AND not defined here, it should assume the value of `default`
+    # to be consistent with how Kubernetes treats resources (i.e., creating a resource
+    # with no metadata.namespace field implies the default namespace) (https://github.com/kyverno/kyverno/issues/4059)
+    # (https://github.com/kyverno/kyverno/issues/4087)
+    namespace: foonamespace # optional
+    kind: Kind # required. Should support values in kind and GVK formats.
+    result: myresult # required
+  # mutate
+  - policy: nextname
+    rule: mutaterulename
+    resources:
+    - resource01
+    patchedResource: mypatchedresource.yaml
+    kind: Kind
+    result: myresult
+  # generate
+  - policy: nextname
+    rule: generaterulename
+    resources:
+    - resource01
+    generatedResource: mypatchedresource.yaml
+    kind: Kind
+    result: myresult
+  # verifyImages. Must support attestations (https://github.com/kyverno/kyverno/issues/2866)
+  - policy: verify-attestations01
+    rule: verifyimagesrulename
+    resources:
+    - resource01
+    kind: Kind
+    result: myresult
+  variables:
```

Comment:
need complete support for Foreach policies variables.

Reply:
I don't have policy but if we have multiple list in foreach policy and both list have variables then right now it's not supported in CLI.

Reply:
Also for multiple list have, one precondition per each also,
we need proper design to support variables in foreach policies by evaluating all scenarios.

Reply:
>Also for multiple list have, one precondition per each also,

I don't understand the above comment.

Reply:
```
foreach:
      - list: "request.object.spec.containers"
        preconditions:
          all:
          - key: "{{ element.resources.limits.memory || '' }}"
            operator: NotEquals
            value: ""
          - key: "{{ element.resources.requests.memory || '' }}"
            operator: NotEquals
            value: ""
        deny:
          conditions:
            any:
            - key: "{{ divide('{{ element.resources.limits.memory }}', '{{ element.resources.requests.memory }}') }}"
              operator: GreaterThan
              value: 2.5
      - list: "request.object.spec.initContainers"
        preconditions:
          all:
          - key: "{{ element.resources.limits.memory || '' }}"
            operator: NotEquals
            value: ""
          - key: "{{ element.resources.requests.memory || '' }}"
            operator: NotEquals
            value: ""
        deny:
          conditions:
            any:
            - key: "{{ divide('{{ element.resources.limits.memory }}', '{{ element.resources.requests.memory }}') }}"
              operator: GreaterThan
              value: 2.5
```
These type of multiple list policies variables are not supported completely

Reply:
Ok, in an example like this, I see these as things that the CLI should be able to harvest from the resource itself right now yet it doesn't. That's what it does today for other `request.object.*` variables and there's no reason it couldn't or shouldn't in cases of `foreach`. The purpose of this KDP and, specifically, what I'm trying to design for in the schema example, is how to cover for things which either aren't possible today because we need structures to represent them, or we have identified improvements to existing ones after years of experience.

Reply:
Marking resolved.